hivelocity/instances_v2.go

```diff
@@ -0,0 +1,163 @@
+	corev1 "k8s.io/api/core/v1"
+	cloudprovider "k8s.io/cloud-provider"
+)
+
+// HVInstancesV2 implements cloudprovider.InstanceV2.
+type HVInstancesV2 struct {
+	client client.Interface
+}
+
+var _ cloudprovider.InstancesV2 = &HVInstancesV2{}
+
+// NewHVInstanceV2 creates a new HVInstancesV2 struct.
+func NewHVInstanceV2(c client.Interface) *HVInstancesV2 {
+	return &HVInstancesV2{client: c}
+}
+
+// getHivelocityDeviceIDFromNode returns the deviceID from a Node.
+// Example: If Node.Spec.ProviderID is "hivelocity://123", then 123
+// will be returned.
+func getHivelocityDeviceIDFromNode(node *corev1.Node) (int32, error) {
+	providerPrefix := providerName + "://"
+	if !strings.HasPrefix(node.Spec.ProviderID, providerPrefix) {
+		return 0, fmt.Errorf("node: %s, ProviderID: %q: %w", node.GetName(), node.Spec.ProviderID, errMissingProviderPrefix)
+	}
+	deviceID, err := strconv.ParseInt(strings.TrimPrefix(node.Spec.ProviderID, providerPrefix), 10, 32)
+	if err != nil {
+		return 0, fmt.Errorf("node: %s, ProviderID %q: %w",
+			node.GetName(), node.Spec.ProviderID, errFailedToConvertProviderID)
+	}
+	return int32(deviceID), nil
+}
+
+var (
+	errMissingProviderPrefix     = fmt.Errorf("missing prefix %q in node.Spec.ProviderID", providerName)
+	errFailedToConvertProviderID = fmt.Errorf("failed to convert node.Spec.ProviderID")
+)
+
+// InstanceExists returns true if the instance for the given node exists according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+// Implements cloudprovider.InstancesV2.InstanceExists.
+func (i2 *HVInstancesV2) InstanceExists(ctx context.Context, node *corev1.Node) (bool, error) {
+	if node == nil {
+		return false, ErrNodeIsNil
+	}
+	deviceID, err := getHivelocityDeviceIDFromNode(node)
+	if err != nil {
+		return false, fmt.Errorf("InstanceExists(): getHivelocityDeviceIDFromNode() failed: %w", err)
+	}
+	_, err = i2.client.GetBareMetalDevice(ctx, deviceID)
+	if errors.Is(err, client.ErrNoSuchDevice) {
+		return false, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("InstanceExists(): GetBareMetalDevice() failed. deviceID %d: %w", deviceID, err)
+	}
+	return true, nil
+}
+
+// ErrUnknownPowerStatus .
+var ErrUnknownPowerStatus = errors.New("unknown PowerStatus")
+
+// ErrNodeIsNil .
+var ErrNodeIsNil = errors.New("node is nil")
+
+// InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+// Implements cloudprovider.InstancesV2.InstanceShutdown.
+func (i2 *HVInstancesV2) InstanceShutdown(ctx context.Context, node *corev1.Node) (bool, error) {
+	if node == nil {
+		return false, ErrNodeIsNil
+	}
+	deviceID, err := getHivelocityDeviceIDFromNode(node)
+	if err != nil {
+		return false, fmt.Errorf("InstanceShutdown(): getHivelocityDeviceIDFromNode() failed. node %s: %w",
```

Comment:
Related to "Node" vs "node": I don't have an opinion here. I sligthly tend to use the upper/lower case according to the variable which gets added to the message. Background: If the variable name is long, then I will copy+paste it from the line above into the string. Then it would be additional work to change it to upper-case.

Reply:
Related "Error: %w": I am unsure. I tend to not include the word "Error". In some cases, this code path might not be an error at all. But if a user sees the message and the word "Error" he might think that there is something broken.

Comparison: Imagine you look at the detailed messages while a linux OS boots. If you see "Error", you think that there is something wrong. But maybe the OS just found no braille input device (which mostly gets used by blind people).

Reply:
Mmh. We can continue like this if you want. The "error" will be an error anyway, as we return it as one and as it gets logged somewhere like this. There I don't think it has much benefit to avoid it, but we can also follow this structure you chose. It is not really grammatically correct and might make it harder to find the error message (as there is no keyword, unlike for "node" here, but it is also easy to follow and might have other benefits.